src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -1608,12 +1623,112 @@ private static unsafe void UInt64ToNumber(ulong value, ref NumberBuffer number)
             number.CheckConsistency();
         }
 
-        internal static unsafe string UInt64ToDecStr(ulong value, int digits)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint Int64DivMod1E9(ref ulong value)
         {
-            if (digits < 1)
-                digits = 1;
+            uint rem = (uint)(value % 1_000_000_000);
+            value /= 1_000_000_000;
+            return rem;
+        }
 
-            int bufferLength = Math.Max(digits, FormattingHelpers.CountDigits(value));
+#if TARGET_64BIT
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+#endif
+        internal static unsafe byte* UInt64ToDecChars(byte* bufferEnd, ulong value)
+        {
+#if TARGET_32BIT
+            while ((uint)(value >> 32) != 0)
+            {
+                bufferEnd = UInt32ToDecChars(bufferEnd, Int64DivMod1E9(ref value), 9);
+            }
+            return UInt32ToDecChars(bufferEnd, (uint)value);
+#else
+            do
+            {
+                ulong remainder;
+                (value, remainder) = Math.DivRem(value, 10);
```

Comment:
We use Math.DivRem here and elsewhere but separate % and / operations above in Int64DivMod1E9, and in both cases the divisor is a constant.  Is there a reason for the difference?

Reply:
`DivRem` is more efficient. `Int64DivMod1E9` is a bit of a special case where we are effectively doing both `32-bit = 64-bit % 32-bit` and `64-bit = 64-bit / 32-bit`. 

The former can be optimized on some platforms, but I didn't do any investigation to confirm if the JIT was doing anything special here or not.